### ISSUE TITLE
Spec for ReduceSumSquare is incorrect when noop_with_empty_axes == 1

### ISSUE THREAD

**Issue by RunnerZhong** [onnx/onnx#6103]
# Bug Report

### Is the issue related to model conversion?

No

### Describe the bug
When noop_with_empty_axes == 1 & axes is empty, in ONNX spec, it will return input tensor directly.
But in reference in onnx, it is mismatch. it returned np.square of the input tensor

![image](https://github.com/onnx/onnx/assets/30307463/0540e792-231c-4ae0-94d5-bc2446da6a79)

### Expected behavior
return input tensor directly

### Notes

@copilot please fix exsisting definitions for ops ReduceSumSquare, ReduceLogSum and ReduceLogSumExp.

**Comment by amankshihab** [comment]
https://github.com/onnx/onnx/blob/d6f87121ba256ac6cc4d1da0463c300c278339d2/docs/Changelog.md?plain=1#L22221-L22222

The expected behavior is mentioned here as well.
Can I work on this?

**Comment by justinchuby** [comment]
Absolutely. Contributions are welcome and appreciated

**Comment by gramalingam** [comment]
This is complicated. Agree that there is a mismatch, but is the bug in the specification or implementation?

My personal interpretation is that this is a bug in the specification, not implementation, for the following reason: the attributes serve to define the set of axes being reduced: specifically, it is a flag to allow the empty list to indicate that all axes must be reduced (or that no axes must be reduced). Now, even if zero axes are reduced, it makes sense to compute the square. ReduceSumSquare is not actually a reduction-op: it is a reduction-op Sum applied to the square of the input.

I think the bug was in reusing the ReduceSum documentation for all ops ... it is correct for basic Reduction ops, but not ReduceSumSquare.

Of course, we can test with other backends/implementations (like onnxruntime, or even pytorch/tensorflow etc. IF they have such an option).

**Comment by RunnerZhong** [comment]
I agree with below idea. So maybe we need to modify the spec of  ops like ReduceSumSquare(ReduceLogSum, ReduceLogSumExp).

> This is complicated. Agree that there is a mismatch, but is the bug in the specification or implementation?
> 
> My personal interpretation is that this is a bug in the specification, not implementation, for the following reason: the attributes serve to define the set of axes being reduced: specifically, it is a flag to allow the empty list to indicate that all axes must be reduced (or that no axes must be reduced). Now, even if zero axes are reduced, it makes sense to compute the square. ReduceSumSquare is not actually a reduction-op: it is a reduction-op Sum applied to the square of the input.
> 
> I think the bug was in reusing the ReduceSum documentation for all ops ... it is correct for basic Reduction ops, but not ReduceSumSquare.
> 
> Of course, we can test with other backends/implementations (like onnxruntime, or even pytorch/tensorflow etc. IF they have such an option).

**Comment by fdwr** [comment]
This is definitely a spec bug for [`ReduceSumSquare`](https://onnx.ai/onnx/operators/onnx__ReduceSumSquare.html) and kin, which still say "When axes is empty and this attribute is set to true ... the output tensor would be equivalent to input tensor". If you look at the [functional decomposition](https://fdwr.github.io/MachineLearningOperators/OperatorFormulas.html) of each of these, you see the additional {exp, pow, log, sqrt, ...} operations occur *before* and *after* the reduction.

```
function reduceSumLog(input, axes, keepDimensions)
    return log(reduceSum(input, axes, keepDimensions)) // logₑ(x[0] + x[1] + ... + x[n-1])
endfunction

function reduceExpSumLog(input, axes, keepDimensions)
    return log(reduceSum(exp(input), axes, keepDimensions)) // logₑ(expₑ(x[0]) + expₑ(x[1]) + ... + expₑ(x[n-1]))
endfunction

function reduceSumSquare(input, axes, keepDimensions)
    return reduceSum(pow(X, 2), axes, keepDimensions) // x[0]^2 + x[1]^2 + ... + x[n-1]^2
endfunction

function reduceL1(input, axes, keepDimensions)
    return reduceSum(abs(input), axes, keepDimensions) // abs(x[0]) + abs(x[1]) + ... + abs(x[n-1])
endfunction

function reduceL2(input, axes, keepDimensions)
    return sqrt(reduceSum(pow(X, 2), axes, keepDimensions)) // sqrt(x[0]^2 + x[1]^2 + ... + x[n-1]^2)
endfunction
```

Another way to rationalize it is to ask what result would you get with an input of shape `[1,1,1,1]` (since this flag effectively turns a multaxis operation into hundreds of individual scalar operations), and the answer is that the {pow, exp, log, sqrt} still apply. So [`ReduceSumSquare`](https://onnx.ai/onnx/operators/onnx__ReduceSumSquare.html) *should* say: "When axes is empty and this attribute is set to true ... the output tensor would be equivalent to **the square of** the input tensor"

**Comment by justinchuby** [comment]
Any contributions to correct spec would be appreciated!

**Comment by justinchuby** [comment]
@fdwr would you like to contribute the fix? We have a branch cut by the end of next week and it would be great to include this fix. Copilot doesn’t cover everything apparently. But it seems there is a fix to the function def that may be legitimate

**Comment by fdwr** [comment]
I had trouble getting the Python tests to pass last time (even with a direct clean build of main, no local changes), but I can try again now. Reading the PR description from CoPilot, it's clearly confused, as the current reference behavior for `ReduceSum` is the correct behavior (`pow(X, 2)` : `x[0]^2 + x[1]^2 + ... + x[n-1]^2`), and it's the spec *wording* which is wrong (copy pasta). So it may be faster to just manually intervene than trying to iterate via prompts.

**Comment by justinchuby** [comment]
The PR description was wrong, but I was able to prompt it to update (parts of) the spec. You can navigate to files changed to see if they are useful?

Feel free to also use CI for testing if that’s easier

**Comment by justinchuby** [comment]
@fdwr I merged the copilot fixes, but I think the wording you have in your branch is better. (also missing was the changes to the ref impl you made). Would you like to create a PR?

**Comment by fdwr** [comment]
> [@fdwr](https://github.com/fdwr) I merged the copilot fixes, but I think the wording you have in your branch is better. (also missing was the changes to the ref impl you made). Would you like to create a PR?

@justinchuby Your wording is fine :). Vote to close as completed by https://github.com/onnx/onnx/pull/7137.

**Comment by justinchuby** [comment]
Reading my previous comment: looks like we still need a change to the reference implementation

**Comment by justinchuby** [comment]
@titaiwangms would you be able to help? Thanks.

**Comment by titaiwangms** [comment]
OK I will take a look.
